pimd/pim6_mroute.c

```diff
@@ -0,0 +1,1120 @@
+		zlog_warn("Failure closing mroute socket: fd=%d errno=%d: %s",
+			  pim->mroute_socket, errno, safe_strerror(errno));
+		return -3;
+	}
+
+	mroute_read_off(pim);
+	pim->mroute_socket = -1;
+
+	return 0;
+}
+
+/*
+  For each network interface (e.g., physical or a virtual tunnel) that
+  would be used for multicast forwarding, a corresponding multicast
+  interface must be added to the kernel.
+ */
+int pim_mroute_add_if(struct interface *ifp, pim_addr ifaddr,
+		       unsigned char flags)
+{
+	struct pim_interface *pim_ifp = ifp->info;
+	struct mif6ctl vc;
+	int err;
+
+	if (PIM_DEBUG_MROUTE)
+		zlog_debug("%s: Add Vif %d (%s[%s])", __func__,
+			   pim_ifp->mroute_if_index, ifp->name,
+			   pim_ifp->pim->vrf->name);
+
+	memset(&vc, 0, sizeof(vc));
+	vc.mif6c_mifi = pim_ifp->mroute_if_index;
+	vc.mif6c_flags = flags;
+	vc.vifc_threshold = PIM_MROUTE_MIN_TTL;
+	vc.vifc_rate_limit = 0;
+
+	err = setsockopt(pim_ifp->pim->mroute_socket, IPPROTO_IPV6, MRT6_ADD_MIF,
+			 (void *)&vc, sizeof(vc));
+	if (err) {
+		zlog_warn(
+			"%s: failure: setsockopt(fd=%d,IPPROTO_IPV6,MRT6_ADD_MIF,mif_index=%d,ifaddr=%pPAs,flag=%d): errno=%d: %s",
+			__func__, pim_ifp->pim->mroute_socket, ifp->ifindex,
+			&ifaddr, flags, errno, safe_strerror(errno));
+		return -2;
+	}
+
+	return 0;
+}
+
+int pim_mroute_del_if(struct interface *ifp)
+{
+	struct pim_interface *pim_ifp = ifp->info;
+	struct mif6ctl vc;
+	int err;
+
+	if (PIM_DEBUG_MROUTE)
+		zlog_debug("%s: Del Mif %d (%s[%s])", __func__,
+			   pim_ifp->mroute_if_index, ifp->name,
+			   pim_ifp->pim->vrf->name);
+
+	memset(&vc, 0, sizeof(vc));
+	vc.mif6c_mifi = pim_ifp->mroute_if_index;
+
+	err = setsockopt(pim_ifp->pim->mroute_socket, IPPROTO_IPV6, MRT6_DEL_MIF,
+			 (void *)&vc, sizeof(vc));
+	if (err) {
+		zlog_warn(
+			"%s %s: failure: setsockopt(fd=%d,IPPROTO_IPV6,MRT6_DEL_MIF,mif_index=%d): errno=%d: %s",
+			__FILE__, __func__, pim_ifp->pim->mroute_socket,
+			pim_ifp->mroute_if_index, errno, safe_strerror(errno));
+		return -2;
+	}
+
+	return 0;
+}
+
+/*
+ * Prevent creating MFC entry with OIF=IIF.
+ *
+ * This is a protection against implementation mistakes.
+ *
+ * PIM protocol implicitely ensures loopfree multicast topology.
+ *
+ * IGMP must be protected against adding looped MFC entries created
+ * by both source and receiver attached to the same interface. See
+ * TODO T22.
+ * We shall allow igmp to create upstream when it is DR for the intf.
+ * Assume RP reachable via non DR.
+ */
+bool pim_mroute_allow_iif_in_oil(struct channel_oil *c_oil,
+		int oif_index)
+{
+#ifdef PIM_ENFORCE_LOOPFREE_MFC
+	struct interface *ifp_out;
+	struct pim_interface *pim_ifp;
+
+	if (c_oil->up &&
+		PIM_UPSTREAM_FLAG_TEST_ALLOW_IIF_IN_OIL(c_oil->up->flags))
+		return true;
+
+	ifp_out = pim_if_find_by_mcast_if_index(c_oil->pim, oif_index);
+	if (!ifp_out)
+		return false;
+	pim_ifp = ifp_out->info;
+	if (!pim_ifp)
+		return false;
+	if ((c_oil->oif_flags[oif_index] & PIM_OIF_FLAG_PROTO_IGMP) &&
```

Comment:
Why PIM_OIF_FLAG_PROTO_IGMP?

Reply:
These flags would be renamed when MLD is introduced. For now this refers to an GM introduced OIF.